### PR TITLE
Time-Warped Nonnegative Matrix Factorization

### PR DESCRIPTION
**Pull request by ahwillia**
This is a prototype for time-warped nonnegative matrix factorization. The model appear more difficult to optimize, so the demo is not currently working.

Also implements truncated SVD for factor initialization.

### CODE REVIEW

File: twpca/utils.py [hunk 1]
```diff
@@ -69,14 +70,11 @@ def compute_lowrank_factors(data, n_components, fit_trial_factors, last_idx, sca
...
-    sqs = np.sqrt(s) * scale
-
-    # set neuron and time factors to top singular vectors
-    time_fctr = (u[:, :n_components] * sqs[:n_components]).astype(np.float32)
-    neuron_fctr = (v[:n_components].T * sqs[:n_components]).astype(np.float32)
+    # do matrix decomposition on trial-averaged data matrix
+    DecompModel = NMF if nonneg else TruncatedSVD
+    model = DecompModel(n_components=n_components)
+    time_fctr = model.fit_transform(np.nanmean(data, axis=0)).astype(np.float32)
+    neuron_fctr = np.transpose(model.components_).astype(np.float32)
```
Comment (poolio) [inline twpca/utils.py]: what about the scalnig of factors?

File: twpca/model.py [hunk 2]
```diff
@@ -52,6 +55,7 @@ def __init__(self, n_components, shared_length=None,
         }
 
         # store tensorflow variables (parameters) and session
+        self._raw_params = {}
```
Comment (poolio) [inline twpca/model.py]: add comment for the raw_param/param distinction

File: twpca/utils.py (hunk 1 above)
Comment (ahwillia) [reply twpca/utils.py]: Probably not important, but is addressed in the next commit to be consistent with previous code.

**Comment by ahwillia** [comment]
Correcting the initialization for the softplus transform did the trick. Ready to review and merge @poolio 

image

^^ the nonneg-demo works

**Review by poolio — COMMENTED** [review]
Great work, @ahwillia !

**Pull request merged**